File: golang/2577.minimum-time-to-visit-a-cell-in-a-grid.go
import "container/heap"

type TimerQue [][3]int

func (h TimerQue) Len() int            { return len(h) }
func (h TimerQue) Less(i, j int) bool  { return h[i][0] < h[j][0] }
func (h TimerQue) Swap(i, j int)       { h[i], h[j] = h[j], h[i] }
func (h *TimerQue) Push(v interface{}) { *h = append(*h, v.([3]int)) }
func (h *TimerQue) Pop() interface{}   { a := *h; v := a[len(a)-1]; *h = a[:len(a)-1]; return v }

func minimumTime(grid [][]int) int {
	const INF = int(1e9)

	if grid[0][1] > 1 && grid[1][0] > 1 {
		return -1
	}

	DIRS := [][2]int{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}

	m, n := len(grid), len(grid[0])
	timer := make([][]int, m)
	for i := 0; i < m; i++ {
		timer[i] = make([]int, n)
		for j := 0; j < n; j++ {
			timer[i][j] = INF
		}
	}

	timer[0][0] = 0

	q := make(TimerQue, 0)
	heap.Push(&q, [3]int{0, 0, 0})

	for len(q) > 0 {
		item := heap.Pop(&q).([3]int)
		time, y, x := item[0], item[1], item[2]

		if y == m-1 && x == n-1 {
			return time
		}

		for _, d := range DIRS {
			ny, nx := y+d[0], x+d[1]

			if ny < 0 || ny >= m || nx < 0 || nx >= n {
				continue
			}

			var newTime int
			if time >= grid[ny][nx] {
				newTime = time + 1
			} else {
				newTime = grid[ny][nx] + (grid[ny][nx]-time+1)%2
			}

			if newTime < timer[ny][nx] {
				timer[ny][nx] = newTime
				heap.Push(&q, [3]int{newTime, ny, nx})
			}
		}

	}

	return -1
}

File: python/2577.minimum-time-to-visit-a-cell-in-a-grid.py
import heapq


class Solution:
    def minimumTime(self, grid: list[list[int]]) -> int:
        if grid[0][1] > 1 and grid[1][0] > 1:
            return -1

        m, n = len(grid), len(grid[0])
        DIR = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        INF = 10**9

        start = (0, 0)
        finish = (m - 1, n - 1)

        hq: list[tuple[int, tuple[int, int]]] = [(0, start)]
        timer = [[INF] * n for _ in range(m)]
        timer[0][0] = 0

        while hq:

            time, node = heapq.heappop(hq)

            if node == finish:
                return time

            for dy, dx in DIR:
                y, x = node[0] + dy, node[1] + dx

                if y < 0 or y >= m or x < 0 or x >= n:
                    continue

                if time >= grid[y][x]:
                    new_time = time + 1
                else:
                    new_time = grid[y][x] + (time - grid[y][x] + 1) % 2

                if new_time < timer[y][x]:
                    heapq.heappush(hq, (new_time, (y, x)))
                    timer[y][x] = new_time

        return -1

